### PR TITLE
fix: Memcopy for mmio is bad, do manual copy

### PR DESCRIPTION
**Pull request by phako**
e.g. on ARM64, memcpy will result in more than 1 write to an address, causing multiple triggers to the FPGA for this data to happen

### CODE REVIEW

**Comment by phako** [comment]
This is the "replacement" for #358 which just fixes the memcpy issue

File: device_backends/uio/src/UioAccess.cc [hunk 1]
```diff
@@ -56,10 +56,11 @@ namespace ChimeraTK {
       throw ChimeraTK::logic_error("UIO: Read request exceeds device memory region");
     }
 
-    void* targetAddress = static_cast<uint8_t*>(_deviceUserBase) + address;
-
-    // Is inherently thread-safe
-    std::memcpy(data, targetAddress, sizeInBytes);
+    volatile int32_t* rptr = static_cast<volatile int32_t*>(_deviceUserBase) + address / sizeof(int32_t);
+    while(sizeInBytes >= sizeof(int32_t)) {
+      *data++ = *rptr++;
```
Comment (killenb) [inline device_backends/uio/src/UioAccess.cc:61]: I never know which operator has precedence. I guess this is ```*(data++) = *(rptr++)```?

Comment (killenb) [reply device_backends/uio/src/UioAccess.cc:61]: Please make it explicit.

File: device_backends/uio/src/UioAccess.cc [hunk 2]
```diff
@@ -56,10 +56,11 @@ namespace ChimeraTK {
       throw ChimeraTK::logic_error("UIO: Read request exceeds device memory region");
     }
 
-    void* targetAddress = static_cast<uint8_t*>(_deviceUserBase) + address;
-
-    // Is inherently thread-safe
-    std::memcpy(data, targetAddress, sizeInBytes);
+    volatile int32_t* rptr = static_cast<volatile int32_t*>(_deviceUserBase) + address / sizeof(int32_t);
+    while(sizeInBytes >= sizeof(int32_t)) {
```
Comment (killenb) [inline device_backends/uio/src/UioAccess.cc:60]: There are up to three bytes uncopied. I guess this is not a valid use case with UIO? Do we have a check for it?

File: device_backends/uio/src/UioAccess.cc [hunk 3]
```diff
@@ -73,10 +74,12 @@ namespace ChimeraTK {
     if(address + sizeInBytes > _deviceMemSize) {
       throw ChimeraTK::logic_error("UIO: Write request exceeds device memory region");
     }
-    void* targetAddress = static_cast<uint8_t*>(_deviceUserBase) + address;
 
-    // Is inherently thread-safe
-    std::memcpy(targetAddress, data, sizeInBytes);
+    volatile int32_t* __restrict__ wptr = static_cast<volatile int32_t*>(_deviceUserBase) + address / sizeof(int32_t);
+    while(sizeInBytes >= sizeof(int32_t)) {
```
Comment (killenb) [inline device_backends/uio/src/UioAccess.cc:79]: see above

File: device_backends/uio/src/UioAccess.cc [hunk 4]
```diff
@@ -73,10 +74,12 @@ namespace ChimeraTK {
...
+      *wptr++ = *data++;
```
Comment (killenb) [inline device_backends/uio/src/UioAccess.cc:80]: see above

File: device_backends/uio/src/UioAccess.cc (hunk 1 above)
Comment (phako) [reply device_backends/uio/src/UioAccess.cc:61]: If you do that, it first increments, then assigns, skipping the first entry.

will do
```
*data = *rptr
data++; rptr++
```

File: device_backends/uio/src/UioAccess.cc (hunk 2 above)
Comment (phako) [reply device_backends/uio/src/UioAccess.cc:60]: Good question, for our UIO this is correct, probably not for the generic case?

File: device_backends/uio/src/UioAccess.cc (hunk 3 above)
Comment (phako) [reply device_backends/uio/src/UioAccess.cc:88]: See above

**Comment by killenb** [comment]
> I just thought that this will also be problematic if we do not start on a multiple of 4?

That's a good point. And wasn't this writing in chunks of 32 to prevent memcopy from writing in 64 bit blocks? What happens on a ARM 32 if you write an uneven number of shorts? Ah no, we write the remainder byte by byte.

**Comment by phako** [comment]
We cannot do copy byte-by-byte, that would be 4 writes per int32 register and we would be back to square 1

**Comment by killenb** [comment]
But what do we do with the remaining not multiple of 4 bytes? We have to write them bye by byte, so it is up to 3 writes.

This is special because it assumes we have an underying  32 bit layout, and having those "trigger registers" aligned with it. We somehow need a way to configure that. Or don't we? Somehow the requirements are not clear.

We could also just check that the size is a multiple of 4 bytes and throw if it is not. But this makes the backed less universal.

**Comment by mhier** [comment]
> We could also just check that the size is a multiple of 4 bytes and throw if it is not. But this makes the backed less universal.

You can simply overload `NumericAddressedBackend::minimumTransferAlignment()` to return 4, in which case there will never be done any transfer which is not aligned at integer multiple of 4 bytes. Smaller registers will be accessed by issuing a wider transfer and throwing away bits (resp. read-modify-write for the write direction). This is how we have it for the pcie backend etc.

**Pull request merged**